src/gwt/src/org/rstudio/studio/client/workbench/UserInterfaceHighlighter.java

```diff
@@ -70,19 +91,109 @@ public Element getHighlightElement()
       {
          return highlightedElement_;
       }
-      
+
+      public void clearHandler()
+      {
+         if (handler_ != null)
+            handler_.removeHandler();
+      }
+
+      public void executeCallback()
+      {
+         // This method must be called by a single HighlightPair, but because there can be multiple
+         // pairs per query, we need to check if the callback has already executed before proceeding.
+         if(!highlighter_.getCallbackProcessed(index_))
+         {
+            highlighter_.setCallbackProcessed(index_, true);
+            highlighter_.getServer().executeRCode(callback_, new ServerRequestCallback<String>(){
+
+               @Override
+               public void onResponseReceived(String results)
+               {
+                  // Remove listener from this element and all other elements with the same query
+                  highlighter_.clearEvents();
+               }
+
+               @Override
+               public void onError(ServerError error)
+               {
+                  Debug.logError(error);
+               }
+            });
+         }
+      }
+
+      private HandlerRegistration addListener()
+      {
+         final JavaScriptObject functor = addEventListener(callback_, monitoredElement_);
+
+         return new HandlerRegistration()
+         {
+            public void removeHandler()
+            {
+               invokeFunctor(functor);
+            }
+         };
+      }
+
+      private native JavaScriptObject addEventListener(String code, Element el)/*-{
+         var thiz = this;
+         var callback = function() {
```

Comment:
nit: you should normally wrap JavaScript functions with `$entry()`; see e.g. 

https://github.com/rstudio/rstudio/blob/9ae3c31da14c08233d235b216688510845892a77/src/gwt/src/org/rstudio/studio/client/shiny/ShinyApplicationPresenter.java#L146-L151

Some documentation on why this is done (primarily to ensure function is executed with an appropriate exception handler for GWT):

http://www.gwtproject.org/doc/latest/DevGuideCodingBasicsJSNI.html

Reply:
Thanks for including the documentation! Fixed.